### PR TITLE
Feature/cp assessments

### PR DESCRIPTION
**Pull request by tsubik**
Adding carbon performance assessment

```
bundle exec rake import:all
```

image

### CODE REVIEW

File: app/admin/companies.rb [hunk 1]
```diff
@@ -54,14 +54,16 @@
...
+              'No Management Quality Assessments for this company yet'
+            end
           else
             resource.mq_assessments.latest_first.map do |a|
               panel "Assessement taken on #{a.assessment_date}", class: 'mq_assessment' do
                 attributes_table_for a do
                   row :level, &:status_description_short
                   row :publication_date do
-                    "#{a.publication_date.strftime('%B %Y')}"
+                    a.publication_date.strftime('%B %Y')
```
Comment (kowal) [inline app/admin/companies.rb]: we could use rails date/time formats for this
https://api.rubyonrails.org/classes/Time.html#method-i-to_formatted_s

File: app/models/mq/assessment.rb [hunk 2]
```diff
@@ -15,7 +15,7 @@
 
 module MQ
   class Assessment < ApplicationRecord
-    LEVELS = %w[1 2 3 4 4STAR].freeze
+    LEVELS = %w[0 1 2 3 4 4STAR].freeze
```
Comment (kowal) [inline app/models/mq/assessment.rb]: what does `4STAR` mean here?

File: app/admin/companies.rb [hunk 3]
```diff
@@ -84,6 +86,48 @@
       end
 
       tab :cp_assessments do
+        panel 'Carbon Performance Assessments' do
+          if resource.cp_assessments.empty?
+            div class: 'padding-20' do
+              'No Carbon Performance Assessments for this company yet'
```
Comment (kowal) [inline app/admin/companies.rb]: I think we could wrap resources in some kind of presenter object, which could handle things like:
- placeholders for whole resource/empty fields (`No .. yet`, `N/A`, ..)
- formatting dates

What do you think?

File: app/models/cp/assessment.rb [hunk 4]
```diff
@@ -0,0 +1,15 @@
+module CP
+  class Assessment < ApplicationRecord
+    belongs_to :company
+
+    scope :latest_first, -> { order(assessment_date: :desc) }
+
+    validates_presence_of :publication_date
+
+    def emissions_all_years
```
Comment (kowal) [inline app/models/cp/assessment.rb]: some doc describing what is being returned here would be nice :)

File: app/services/import_companies.rb [hunk 5]
```diff
@@ -39,6 +39,27 @@ def header_converter
...
+    MQ::Assessment.create!(
+      mq_assessment_attributes(row).merge(company_id: company.id)
+    )
+  end
+
+  def create_cp_assessment!(row, company)
+    assumptions = row[:assumptions]&.strip
+    cp_assessment_date = row[:carbon_performance_assessment_date]&.strip&.downcase
+
+    return if cp_assessment_date == 'not assessed' && assumptions.blank?
```
Comment (kowal) [inline app/services/import_companies.rb]: I would extract method:
```rb
def create_cp_assessment!(row, company)
  return if assessment_row_invalid?(row) # not sure about the name
  ..
```

File: app/services/import_companies.rb [hunk 6]
```diff
@@ -76,6 +97,12 @@ def get_questions(row)
     end.compact
   end
 
+  def get_emissions(row)
+    row.headers.grep(/^\d{4}$/).map do |year|
```
Comment (kowal) [inline app/services/import_companies.rb]: what do you think about reformatting this like this?
```rb
    row.headers
      .grep(/^\d{4}$/)
      .map { |year| { year.to_s.to_i => row[year]&.to_f } }
      .reduce(&:merge)
      .reject { |_k, v| v.blank? }
```

Comment (tsubik) [reply app/services/import_companies.rb]: right :+1: I'm not taking much care into those importers, not sure if they will stay

**Comment by kowal** [comment]
@tsubik we also need to add 'CP Assessments` to main `Assessments` tab, right? Or is it another story?

File: app/admin/companies.rb (hunk 3 above)
Comment (tsubik) [reply app/admin/companies.rb]: Good idea. Can try to do that with some SimpleDelegator

File: app/models/mq/assessment.rb (hunk 2 above)
Comment (tsubik) [reply app/models/mq/assessment.rb]: 4STAR is like 5 level, they named it like this, we can rename it to 5, but then we will have to take care of mapping that back and forth

**Comment by tsubik** [comment]
Yeah, I can add it, not sure if that's needed but it will be like 10 minutes to add.

File: app/services/import_companies.rb (hunk 5 above)
Comment (tsubik) [reply app/services/import_companies.rb]: hmm, actually I'm striping and downcasing values and using them to create assessment later so I wouldn't move that into a separate method to not do it again.

**Comment by tsubik** [comment]
Ready again @kowal

**Comment by tsubik** [comment]
Also, I updated npm dependencies to make github security alert go away.

File: app/services/import_companies.rb (hunk 5 above)
Comment (kowal) [reply app/services/import_companies.rb]: ok, that's fine then

File: app/models/mq/assessment.rb (hunk 2 above)
Comment (kowal) [reply app/models/mq/assessment.rb]: so maybe just add a comment above, cause this looks like some typo ;)

**Pull request merged**